File: frontend/src/api/payment.ts
import { apiRequest } from './client'

export interface Payment {
  payment_id: number
  amount: number
  original_amount: number | null
  discount_applied: number | null
  currency: string
  status: string
  description: string | null
  subscription_duration_months: number | null
  provider: string
  created_at: string
}

export interface PaymentsList {
  items: Payment[]
  total: number
  page: number
  limit: number
}

export interface CreatePaymentRequest {
  provider: string
  months: number
  promo_code?: string
}

export interface CreatePaymentResponse {
  payment_id: number
  redirect_url: string
  amount: number
  original_amount: number | null
  currency: string
}

export interface PaymentStatus {
  payment_id: number
  status: string
  provider: string
  amount: number
  currency: string
  created_at: string
}

export interface PromoApplyResponse {
  promo_code: string
  discount_percentage: number
  expires_at: string
}

export function getPayments(page = 1, limit = 20): Promise<PaymentsList> {
  return apiRequest<PaymentsList>(`/payments?page=${page}&limit=${limit}`)
}

export function getPaymentsCount(): Promise<{ total: number }> {
  return apiRequest<{ total: number }>('/payments/count')
}

export function createPayment(data: CreatePaymentRequest): Promise<CreatePaymentResponse> {
  return apiRequest<CreatePaymentResponse>('/payments/create', {
    method: 'POST',
    body: JSON.stringify(data),
  })
}

export function getPaymentStatus(paymentId: number): Promise<PaymentStatus> {
  return apiRequest<PaymentStatus>(`/payments/${paymentId}/status`)
}

export function applyPromo(code: string): Promise<PromoApplyResponse> {
  return apiRequest<PromoApplyResponse>('/promo/apply', {
    method: 'POST',
    body: JSON.stringify({ code }),
  })
}

export function removeActiveDiscount(): Promise<void> {
  return apiRequest<void>('/promo/active-discount', { method: 'DELETE' })
}

File: web/schemas/payment.py
from typing import Optional
from datetime import datetime
from pydantic import BaseModel, Field, model_validator
from web.schemas.types import UTCDatetime


class PaymentResponse(BaseModel):
    payment_id: int
    amount: float
    original_amount: Optional[float]
    discount_applied: Optional[float]
    currency: str
    status: str
    description: Optional[str]
    subscription_duration_months: Optional[int]
    provider: str
    created_at: UTCDatetime


class PaymentsListResponse(BaseModel):
    items: list[PaymentResponse]
    total: int
    page: int
    limit: int


class CreatePaymentRequest(BaseModel):
    provider: str
    months: Optional[int] = Field(default=None, ge=1, le=120)
    plan_option_id: Optional[int] = Field(default=None, ge=1)
    addon_option_id: Optional[int] = Field(default=None, ge=1)
    promo_code: Optional[str] = None

    @model_validator(mode="after")
    def check_months_or_option(self) -> "CreatePaymentRequest":
        if self.months is None and self.plan_option_id is None:
            raise ValueError("Укажите months или plan_option_id")
        if self.addon_option_id is not None and self.plan_option_id is None:
            raise ValueError("Дополнение можно добавить только к каталожному тарифу")
        return self


class CreatePaymentResponse(BaseModel):
    payment_id: int
    redirect_url: str
    amount: float
    original_amount: Optional[float]
    currency: str


class PaymentsCountResponse(BaseModel):
    total: int


class PaymentStatusResponse(BaseModel):
    payment_id: int
    status: str
    provider: str
    amount: float
    original_amount: Optional[float]
    discount_applied: Optional[float]
    currency: str
    description: Optional[str]
    subscription_duration_months: Optional[int]
    redirect_url: Optional[str]
    promo_code: Optional[str]
    created_at: UTCDatetime
